sshd/auth.go

```diff
@@ -34,11 +47,32 @@ func MakeAuth(auth Auth) *ssh.ServerConfig {
 			}}
 			return perm, nil
 		},
+
+		// We use KeyboardInteractiveCallback instead of PasswordCallback to
+		// avoid preventing the client from including a pubkey in the user
+		// identification.
 		KeyboardInteractiveCallback: func(conn ssh.ConnMetadata, challenge ssh.KeyboardInteractiveChallenge) (*ssh.Permissions, error) {
-			if !auth.AllowAnonymous() {
-				return nil, errors.New("public key authentication required")
+			err := auth.CheckBans(conn.RemoteAddr(), nil, sanitize.Data(string(conn.ClientVersion()), 64))
+			if err != nil {
+				return nil, err
+			}
+			if auth.AcceptPassphrase() {
+				var answers []string
+				answers, err = challenge("", "", []string{"Passphrase required to connect: "}, []bool{true})
+				if err == nil {
+					if len(answers) != 1 {
+						err = errors.New("didn't get passphrase")
+					} else {
+						err = auth.CheckPassphrase(answers[0])
+						if err != nil {
+							// TODO: make rate-limiting configurable
+							auth.BanAddr(conn.RemoteAddr(), time.Minute * 1)
```

Comment:
I don't think we need to ban for a full minute, the idea here is to throttle brute forcing (thousands of requests per second) so just a second or two will be enough and that should resolve the UX concerns you had too.

Ideally we'd want to ban at increasing intervals, but I don't think it's worth storing additional state to achieve this right now.

Reply:
Also I'm down for making the ban duration configurable, at least passed in from the Host or even from `main`. We can add a cli flag later if we want.

Reply:
OK, I was obviously thinking in the wrong order of magnitude. I set it to two seconds. With an interval that short, I agree we don't need any options to change it.